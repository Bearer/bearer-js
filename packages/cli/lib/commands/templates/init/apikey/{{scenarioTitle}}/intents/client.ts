import axios from 'axios'

export const CLIENT = axios.create({
  baseURL: 'https://api.example.com/',
  timeout: 5000
})

export function authorizationHeaderWith(token) {
  const encodedAuth = Buffer.from(`anything:${token}`).toString('base64')

  return {
    Authorization: `Basic ${encodedAuth}`
  }
}