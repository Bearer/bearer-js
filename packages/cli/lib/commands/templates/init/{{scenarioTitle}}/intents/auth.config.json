{
  "authorizationURL": null ,
  "tokenURL": null,
  "clientID": null,
  "clientSecret": null,
  "config": {
    "scope": []
  }
}
