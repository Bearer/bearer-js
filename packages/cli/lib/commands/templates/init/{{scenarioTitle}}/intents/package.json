{
  "name": "{{scenarioTitle}}-intents",
  "version": "0.0.1",
  "description": "{{scenarioTitle}} intents",
  "main": "index.js",
  "dependencies": {
    "@bearer/intents": "latest",
    "axios": "^0.18.0"
  },
  "devDependencies": {
    "@bearer/bearer-cli": "latest",
    "@types/node": "latest"
  },
  "license": "MIT"
}
