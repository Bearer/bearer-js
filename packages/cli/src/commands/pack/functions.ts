import * as archiver from 'archiver'
import * as fs from 'fs-extra'
import * as path from 'path'

import BaseCommand from '../../base-command'
import { RequireIntegrationFolder } from '../../utils/decorators'
import prepareConfig from '../../utils/prepare-config'

const CONFIG_FILE = 'bearer.config.json'
const HANDLER_NAME = 'index'
const HANDLER_NAME_WITH_EXT = [HANDLER_NAME, 'js'].join('.')

export default class PackFunctions extends BaseCommand {
  static description = 'Pack integration functions'
  static hidden = true
  static flags = {
    ...BaseCommand.flags
  }

  static args = [{ name: 'ARCHIVE_PATH', required: true }]

  @RequireIntegrationFolder()
  async run() {
    // we are assuming prepare and build have been run previously
    const { args } = this.parse(PackFunctions)
    const { ARCHIVE_PATH } = args
    const target = path.resolve(ARCHIVE_PATH)
    this.debug(`Zipping to: ${target}`)

    const output = fs.createWriteStream(target)
    const archive = archiver('zip', {
      zlib: { level: 9 }
    })
    const archiveEntries: archiver.EntryData[] = []
    archive.pipe(output)

    const archived = new Promise<archiver.EntryData[]>((resolve, reject) => {
      output.on('close', () => {
        resolve(archiveEntries)
      })
      archive.on('error', (err: any) => {
        reject(err)
      })

      archive.on('entry', (entry: archiver.EntryData) => {
        if (!entry.stats || !entry.stats.isDirectory()) {
          archiveEntries.push(entry)
        }
      })

      archive.on('warning', (err: any) => {
        this.debug(err)
      })
    })

    try {
      // add functions
      archive.directory(this.locator.buildFunctionsDir, false)
      // add CONFIG
      const functions = await this.retrieveFunctions()

      const { config, handlers } = buildLambdaDefinitions(functions)
      const finalConfig = await this.retrieveAuthConfig(config)

      archive.append(JSON.stringify(finalConfig, null, 2), { name: CONFIG_FILE })

      // add handlers
      this.debug(`Generated config: ${JSON.stringify(config, null, 2)}`)
      archive.append(handlers, { name: HANDLER_NAME_WITH_EXT })

      // ZIP
      archive.finalize()

      const entries = await archived

      this.debug(`Zip content: ${entries.length} files\n  * ${entries.map(e => e.name).join('\n  * ')}`)
      this.success('Successfully generated lambda package')
      // log files added to zip
    } catch (e) {
      this.error(e)
    }
  }

  async retrieveAuthConfig(config: { functions: Record<string, string>[] }): Promise<TBearerConfig> {
    // generate config
    const content = await fs.readFile(this.locator.authConfigPath, { encoding: 'utf8' })
    return {
      ...config,
      auth: JSON.parse(content)
    }
  }

  // TODO: rewrite this using TS AST
  async retrieveFunctions(): Promise<TFunctionNames> {
    try {
      const config = await prepareConfig(
        this.locator.authConfigPath,
        this.bearerConfig.integrationUuid,
        this.locator.srcFunctionsDir
      )
      return config.functions
    } catch (e) {
      throw e
    }
  }
}

type TFunctionNames = string[]

type TBearerConfig = {
  functions: Record<string, string>[]
  auth?: any
}

type TLambdaDefinition = {
  config: {
    functions: Record<string, string>[]
  }
  handlers: string
}

export function buildLambdaDefinitions(functions: TFunctionNames): TLambdaDefinition {
  return {
    config: {
      functions: functions.reduce(
        (acc, functionName) => {
          acc.push({ [functionName]: [HANDLER_NAME, functionName].join('.') })
          return acc
        },
        [] as Record<string, string>[]
      )
    },
    handlers: buildLambdaIndex(functions)
  }
}

function buildLambdaIndex(functions: TFunctionNames): string {
  return functions
    .map((func, index) => {
      const funcConstName = `func${index}`
      return `const ${funcConstName} = require("./dist/${func}").default;
module.exports['${func}'] = ${funcConstName}.init();
`
    })
    .join('\n')
}
