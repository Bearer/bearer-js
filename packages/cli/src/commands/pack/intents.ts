import * as archiver from 'archiver'
import * as fs from 'fs-extra'
import * as path from 'path'

import BaseCommand from '../../base-command'
import { RequireScenarioFolder } from '../../utils/decorators'
import prepareConfig from '../../utils/prepare-config'

const CONFIG_FILE = 'bearer.config.json'
const HANDLER_NAME = 'index'
const HANDLER_NAME_WITH_EXT = [HANDLER_NAME, 'js'].join('.')

export default class PackIntents extends BaseCommand {
  static description = 'Pack scenario intents'
  static hidden = true
  static flags = {
    ...BaseCommand.flags
  }

  static args = [{ name: 'ARCHIVE_PATH', required: true }]

  @RequireScenarioFolder()
  async run() {
    // we are assuming prepare and build have been run previously
    const { args } = this.parse(PackIntents)
    const { ARCHIVE_PATH } = args
    const target = path.resolve(ARCHIVE_PATH)
    this.debug(`Zipping to: ${target}`)

    const output = fs.createWriteStream(target)
    const archive = archiver('zip', {
      zlib: { level: 9 }
    })
    const archiveEntries: Array<archiver.EntryData> = []
    archive.pipe(output)

    const archived = new Promise<Array<archiver.EntryData>>((resolve, reject) => {
      output.on('close', () => {
        resolve(archiveEntries)
      })
      archive.on('error', (err: any) => {
        reject(err)
      })

      archive.on('entry', (entry: archiver.EntryData) => {
        if (!entry.stats || !entry.stats.isDirectory()) {
          archiveEntries.push(entry)
        }
      })

      archive.on('warning', (err: any) => {
        this.debug(err)
      })
    })

    try {
      // add intents
      archive.directory(this.locator.buildIntentsDir, false)
      // add CONFIG
      const intents = await this.retrieveIntents()

      const { config, handlers } = buildLambdaDefinitions(intents)
      const finalConfig = await this.retrieveAuthConfig(config)

      archive.append(JSON.stringify(finalConfig, null, 2), { name: CONFIG_FILE })

      // add handlers
      this.debug(`Generated config: ${JSON.stringify(config, null, 2)}`)
      archive.append(handlers, { name: HANDLER_NAME_WITH_EXT })

      // ZIP
      archive.finalize()

      const entries = await archived

      this.debug(`Zip content: ${entries.length} files\n  * ${entries.map(e => e.name).join('\n  * ')}`)
      this.success('Successfully generated lambda package')
      // log files added to zip
    } catch (e) {
      this.error(e)
    }
  }

  async retrieveAuthConfig(config: { intents: Record<string, string>[] }): Promise<TBearerConfig> {
    // generate config
    const content = await fs.readFile(this.locator.authConfigPath, { encoding: 'utf8' })
    return {
      ...config,
      auth: JSON.parse(content)
    }
  }

  // TODO: rewrite this using TS AST
  async retrieveIntents(): Promise<TIntentNames> {
    try {
      const config = await prepareConfig(
        this.locator.authConfigPath,
        this.bearerConfig.scenarioUuid,
        this.locator.srcIntentsDir
      )
      return config.intents
    } catch (e) {
      throw e
    }
  }
}

type TIntentNames = string[]

type TBearerConfig = {
  intents: Record<string, string>[]
  auth?: any
}

type TLambdaDefinition = {
  config: {
    intents: Record<string, string>[]
  }
  handlers: string
}

export function buildLambdaDefinitions(intents: TIntentNames): TLambdaDefinition {
  return {
    config: {
      intents: intents.reduce(
        (acc, intentName) => {
          acc.push({ [intentName]: [HANDLER_NAME, intentName].join('.') })
          return acc
        },
        [] as Record<string, string>[]
      )
    },
    handlers: buildLambdaIndex(intents)
  }
}

function buildLambdaIndex(intents: TIntentNames): string {
  return intents
    .map((intent, index) => {
      const intentConstName = `intent${index}`
      const intentInstance = `${intentConstName}Instance`
      return `const ${intentConstName} = require("./dist/${intent}").default;
const ${intentInstance} = new ${intentConstName}()
module.exports['${intent}'] = ${intentInstance}.intentType.intent(${intentInstance}.action);
`
    })
    .join('\n')
}
