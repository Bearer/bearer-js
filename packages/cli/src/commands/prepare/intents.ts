import BaseCommand from '../../base-command'
import { RequireIntegrationFolder } from '../../utils/decorators'
import { ensureFolderExists } from '../../utils/helpers'

export default class PrepareIntents extends BaseCommand {
  static description = 'describe the command here'
  static hidden = true
  static flags = {
    ...BaseCommand.flags
  }

  static args = []

  @RequireIntegrationFolder()
  async run() {
    ensureFolderExists(this.locator.buildIntentsDir, true)
    ensureFolderExists(this.locator.buildArtifactDir, true)
  }
}
