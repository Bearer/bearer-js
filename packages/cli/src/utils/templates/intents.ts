import { Authentications } from '@bearer/types/lib/authentications';
import IntentType from '@bearer/types/lib/intent-types';
import * as Case from 'case';
import * as path from 'path';

import {copyFiles} from '../helpers'
import baseCommand from '../../base-command';


export default async (command: baseCommand, auth: Authentications, type: IntentType, name: string ) => {
  const vars = getVars(name,  auth, type)
  return await copyFiles(command, path.join(`generate/intent`, auth, type), command.locator.srcIntentsDir, vars)
}

function getVars(name: string, authType: Authentications, intentType: IntentType,) {
  return {
    authType,
    intentType,
    fileName: name,
    intentClassName: Case.pascal(name)
  }
}