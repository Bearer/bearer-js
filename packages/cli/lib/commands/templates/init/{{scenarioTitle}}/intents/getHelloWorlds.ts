import { GetCollection } from '@bearer/intents'
import { CLIENT } from './client'

export default class GetHelloWorlsIntent {
  static intentName: string = 'getHelloWorlds'
  static intentType: any = GetCollection

  static action(token: string, params: any, callback: (params: any) => void) {
    CLIENT.get('/people').then(({ data }) => {
      console.log(data)
      callback({ collection: data.results })
    })
  }
}
