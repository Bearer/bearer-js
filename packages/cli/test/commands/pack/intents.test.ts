import { buildLambdaDefinitions } from '../../../src/commands/pack/intents'

describe('pack:intents', () => {
  describe('lambdas declaration', () => {
    const { config, handlers } = buildLambdaDefinitions(['my-intent', 'spongeBobIntent'])

    it('generates a correct handlers', () => {
      expect(handlers).toMatchSnapshot()
    })

    it('build intents bearer config', () => {
      expect(config).toEqual({
        intents: [
          {
            'my-intent': 'index.my-intent'
          },
          {
            spongeBobIntent: 'index.spongeBobIntent'
          }
        ]
      })
    })
  })
})
