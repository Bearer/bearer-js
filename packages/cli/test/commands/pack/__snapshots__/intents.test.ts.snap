// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pack:intents lambdas declaration generates a correct handlers 1`] = `
"const intent0 = require(\\"./dist/my-intent\\").default;
const intent0Instance = new intent0()
module.exports['my-intent'] = intent0Instance.intentType.intent(intent0Instance.action);

const intent1 = require(\\"./dist/spongeBobIntent\\").default;
const intent1Instance = new intent1()
module.exports['spongeBobIntent'] = intent1Instance.intentType.intent(intent1Instance.action);
"
`;
